### DISCUSSION TITLE
TCP Connection issues on iOS when app wakes up

### DISCUSSION THREAD

**Discussion by gordongrech** [baresip/re#1214] [Q&A] [answered]
Hello, 

I am building an app with baresip and re on iOS. The app works very well, however I am experiencing an easily reproducible issue as follows: 

- When the app is in background (Not terminated) 
- And the phone is idle for a while (approx. 2 minutes)
- And an incoming call is received
- I get flooded with these errors:  "tcp: recv handler: recv(): Socket is not connected [57]"

I've traced this down to src/tcp/tcp.c line 400.

I am using baresip and re 3.16 and I have the barsip loop running continuously. 

To confirm the issue relates to some "stale" connection I have tried to start and stop baresip with each call and can confirm that the issue does not happen (I did this just for testing. In my case this approach is not desirable as it introduces a lot of added complexity and overhead).

Thank you for your support.

**Comment by sreimers** [comment, marked as answer]
I think, this is nothing a library itself can handle. The recommended way is to listen for `applicationDidEnterBackground` and close all open connections.

> Once your app goes into the background, it may be suspended. Once it is suspended, it's unable to properly process incoming connections on the listening socket. However, the socket is still active as far as the kernel is concerned. If a client connects to the socket, the kernel will accept the connection but your app won't communicate over it. Eventually the client will give up, but that might take a while.
Thus, it's better to close the listening socket when going into the background, which will cause incoming connections to be immediately rejected by the kernel.

https://developer.apple.com/library/archive/technotes/tn2277/_index.html

**Reply by gordongrech** [reply]
Thank you very much for your swift response and assistance. This makes a lot of sense. I will work a solution over the coming days and will keep this discussion updated accordingly.

**Reply by gordongrech** [reply]
I have been testing this for a while now and making sure baresip is closed whenever it is not in use solves this issue. 

I had to rework my app to start baresip (and re) when needed and close the service again when not in use. So far so good. Thanks again for your support.
